Paxel Type Inferencer

cromwellian opened the pull request:
Will be used for the following:
* Checking Paxel Expressions against HandleConnection types before evaluating
* Linting/Suggesting tighter bounds on inputs
* Allowing some `GlobalFunction` expressions to know their expected return type

Theory of operation: A form of abstract interpretation
* `InferredType` hierarchy represents types
* Literals become `InferredType.Primitive`
* Mathematical operations on compatible types widen to the biggest type that encloses operands
* Unary/Binary ops produce Numeric or Boolean types
* Functions may produce `UnionType` as can the `IfNull` operator
* QualifiedExpressions are `SeqType<ScopeType>` -> `SeqType<ScopeType>`

* Populate a scope with the types of variables, instead of their values
* Start at bottom of tree and propagate types upward
* Each node is "evaluated" similar to how Evaluator would, but without iteration over sequences
* `InferredType.isAssignableFrom()` allows one to check if the resulting type is assignment compatible with an expected type

** Followup:
** Wire up into `EvaluatorParticle`

cromwellian commented:
/gcbrun

cromwellian commented:
/gcbrun

alxrsngrtn commented on java/arcs/core/data/expression/Builders.kt:
```diff
@@ -213,6 +213,31 @@ open class MapScope<V>(
 
   override fun toString() = map.toString()
+
+  private val allKeys get() = map.keys + (parentScope?.map?.keys ?: emptySet())
```
What's the type of this? I find it unclear.

cromwellian replied:
its a set

alxrsngrtn commented on java/arcs/core/data/expression/Builders.kt:
```diff
@@ -213,6 +213,31 @@ open class MapScope<V>(
+
+  private val allKeys get() = map.keys + (parentScope?.map?.keys ?: emptySet())
+
+  override fun equals(other: Any?): Boolean {
```
Maybe, null check first? Was this code-generated?

alxrsngrtn commented on java/arcs/core/data/expression/Evaluator.kt:
```diff
@@ -164,42 +168,69 @@ sealed class GlobalFunction(val name: String) {
         toSequence<Any>(args[1]).asIterable()
       ).asSequence()
+
+    override fun inferredType(evaluator: TypeEvaluator, args: List<InferredType>) =
```
Can this fit on one line? If not, then it should be wrapped with `{}` (according to the recent Kt tech talk).

alxrsngrtn replied:
This style, if accepted, should apply to the rest of the file.

alxrsngrtn commented on java/arcs/core/data/expression/Evaluator.kt:
```diff
@@ -164,42 +168,69 @@ sealed class GlobalFunction(val name: String) {
       toSequence<Comparable<Comparable<*>>>(args[0]).max()
+
+    override fun inferredType(evaluator: TypeEvaluator, args: List<InferredType>) =
+      widen(*args.toTypedArray())
```
... e.g.

```
override fun inferredType(evaluator: TypeEvaluator, args: List<InferredType>) = widen(
   *args.toTypedArray()
)
```

alxrsngrtn commented on java/arcs/core/data/expression/Evaluator.kt:
```diff
@@ -164,42 +168,69 @@ sealed class GlobalFunction(val name: String) {
+    override fun inferredType(evaluator: TypeEvaluator, args: List<InferredType>) =
+      InferredType.UnionType(
+        InferredType.Primitive.NullType,
+        when (val it = args[0]) {
```
`it` is not that descriptive of a name.

alxrsngrtn replied:
What if `args` is an empty list?

alxrsngrtn replied:
If it's two cases, maybe an `if` statement makes sense.

cromwellian replied:
Its not allowed to take zero args and the parser will enforce that at some point in the future. The TS parser already has the arity of functions encoded, but it's not used.

alxrsngrtn commented on java/arcs/core/data/expression/InferredType.kt:
```diff
@@ -0,0 +1,123 @@
+    }
+  }
+
+  for (type in widenOrder) {
```
I think `first` with a predicate would be more appropriate here. Although, you'd have to do some wrangling to still throw the illegal arg exception.

cromwellian replied:
Good suggestion, I think `firstOrNull() ?: throw ... ` will allow the exception to be handled.

alxrsngrtn commented on java/arcs/core/data/expression/InferredType.kt:
```diff
@@ -0,0 +1,123 @@
+
+/** Widen a list of types to the largest type that encompasses all of them. */
+fun widen(vararg types: InferredType): InferredType.Primitive {
+  types.forEach {
```
Hmm... why `forEach` here and `for` down there? 

This reminds me a note Maria told me about from the TS codebase: even though `forEach` is an option, `for` is more readable.

cromwellian replied:
You can early-exit from for() you can't return from forEach{}. In general, we've been preferring functional operators as a pattern in Kotlin. Not sure what readability thinks.

alxrsngrtn commented on java/arcs/core/data/expression/InferredType.kt:
```diff
@@ -0,0 +1,123 @@
+ * may output.
+ */
+sealed class InferredType {
+  open fun isAssignableFrom(other: InferredType) = this == other
```
I think adding `Boolean` would make this slightly more readable.

alxrsngrtn replied:
... as a type for the fn.

alxrsngrtn commented on java/arcs/core/data/expression/InferredType.kt:
```diff
@@ -0,0 +1,123 @@
+      }
+
+      /** Given a [KClass] return the corresponding [InferredType]. */
+      fun of(kClass: KClass<*>) = kclassToType[kClass]!!
```
`require` over `!!`, please.

alxrsngrtn replied:
Is this here just to get the types to agree? If so, there might be another way to go about it.

alxrsngrtn commented on java/arcs/core/data/expression/InferredType.kt:
```diff
@@ -0,0 +1,123 @@
+  data class UnionType(val types: Set<InferredType>) : InferredType() {
+    constructor(vararg types: InferredType) : this(setOf(*types))
+
+    override fun isAssignableFrom(other: InferredType): Boolean = when (other) {
```
Seems like `if` is more appropriate.

alxrsngrtn commented on java/arcs/core/data/expression/TypeEvaluator.kt:
```diff
@@ -0,0 +1,155 @@
+      )
+
+      // Special ops
+      Expression.BinaryOp.IfNull -> expr.left.accept(this, ctx).union(expr.right.accept(this, ctx))
```
Might be nice to call union as an `infix` operator.

alxrsngrtn commented on java/arcs/core/data/expression/TypeEvaluator.kt:
```diff
@@ -0,0 +1,155 @@
+    InferredType.Primitive.NullType
+
+  private fun seqToScopeOr(inferredType: InferredType?, ctx: Scope): Scope =
+    ((inferredType as? SeqType)?.type as? ScopeType)?.scope ?: ctx
```
Are you sure you can't fit more `?` in here?

alxrsngrtn replied:
:)

alxrsngrtn commented on java/arcs/core/data/expression/TypeEvaluator.kt:
```diff
@@ -0,0 +1,155 @@
+  private fun seqToScopeOr(inferredType: InferredType?, ctx: Scope): Scope =
+    ((inferredType as? SeqType)?.type as? ScopeType)?.scope ?: ctx
+
+  private fun InferredType.asScope() = (this as SeqType).type as ScopeType
```
I think ext functions should be at the top level, not in the class. Esp. if it's private. WDYT?

piotrswigon replied:
The method name could be better. When you know that x is seq and you see `x.asScope()` it seems wrong. Maybe `x.getSequenceScope()` ? A bit verbose I know...

alxrsngrtn commented on java/arcs/core/data/expression/TypeEvaluator.kt:
```diff
@@ -0,0 +1,155 @@
+  }
+
+  override fun <T> visit(expr: Expression.FunctionExpression<T>, ctx: Scope): InferredType {
+    val arguments = expr.arguments.map { it.accept(this, ctx) }.toList()
```
I think `.toList` may be redundant.

alxrsngrtn commented on javatests/arcs/core/data/expression/TypeCheckerTest.kt:
```diff
@@ -0,0 +1,300 @@
+    checkAssignableTo(true.asExpr() and true.asExpr(), BooleanType)
+    checkAssignableTo(true.asExpr() or true.asExpr(), BooleanType)
+
+    checkAssignableTo(1.asExpr() ifNull "hello".asExpr(), UnionType(NumberType, TextType))
```
Are there other cases besides `ifNull` that will create union types? I'd like to see more tests for those.

cromwellian replied:
ifNull creates union types, and the union() function creates union types, as does first(). Any input handles with union types or nullable types will also create union types, since that is how we represent nullable.

piotrswigon commented on java/arcs/core/data/expression/TypeEvaluator.kt:
```diff
@@ -0,0 +1,155 @@
+    } else {
+      (expr.qualifier.accept(this, ctx) as? ScopeType)?.scope
+    }).apply {
+      @Suppress("SENSELESS_COMPARISON") if (this == null && !expr.nullSafe) {
```
I think SENSELESS_COMPARISON is telling us that `this` cannot be null.

piotrswigon replied:
nullSafe should carry over the nullability from the type of the qualifier to the type of the looked up field.

piotrswigon commented on java/arcs/core/data/expression/TypeEvaluator.kt:
```diff
@@ -0,0 +1,155 @@
+    InferredType.Primitive.NullType
+
+  private fun seqToScopeOr(inferredType: InferredType?, ctx: Scope): Scope =
+    ((inferredType as? SeqType)?.type as? ScopeType)?.scope ?: ctx
```
Is there a reason to be so lenient with `as?` instead of `as`? I am worried that we could swallow an error (having a qualifier that is not of type Seq, or not `Seq<Scope>` would be an error I think) and don't notice it.

piotrswigon commented on java/arcs/core/data/expression/TypeEvaluator.kt:
```diff
@@ -0,0 +1,155 @@
+  }
+
+  override fun visit(expr: Expression.WhereExpression, ctx: Scope): InferredType {
+    return expr.qualifier.accept(this, ctx)
```
as everywhere else, I would like us to verify types as well, Here I would expect us to evaluate the type of `expr.expr`(chicken chicken!) and ensure it is `Boolean`, wdyt?

piotrswigon commented on java/arcs/core/data/expression/TypeEvaluator.kt:
```diff
@@ -0,0 +1,155 @@
+  }
+
+  override fun <T> visit(expr: Expression.OrderByExpression<T>, ctx: Scope): InferredType {
+    return expr.qualifier.accept(this, ctx)
```
verify that expressions by which we order are... I guess any primitive is allowed, but scopes and sequences are not?

cromwellian replied:
Yes, primitives are allowed. Union is allowed if and only if the types are all primitives. So really, only seqtype and scopetype are banned.

piotrswigon commented on javatests/arcs/core/data/expression/TypeCheckerTest.kt:
```diff
@@ -0,0 +1,300 @@
+
+  @Test
+  fun check_literals() {
+    checkAssignableTo(1.0.asExpr(), DoubleType)
```
you could hypothetically have a bug in assignability checking that always return `true` and all tests would pass. It seems to me it would be better if we also had some `checkNotAssignableTo` mixed in, wdyt? Alternatively, maybe we just need separate tests for assignability logic.

cromwellian replied:
Yeah, i thought of that and tested it manually :) But I'll add negative tests too.

piotrswigon commented on javatests/arcs/core/data/expression/TypeCheckerTest.kt:
```diff
@@ -0,0 +1,300 @@
+  @Test
+  fun check_binary_ops() {
+    checkAssignableTo(1.asExpr() + 2.asExpr(), IntType)
+    checkAssignableTo(1.0.asExpr() + 2.asExpr(), DoubleType)
```
If there was a bug in widening and `1.0.asExpr() + 2.asExpr()` got `IntType` inferred, this assertion would pass. This makes me wonder why you perform all the tests in terms of `checkAssignableTo`, why don't do simple equality here?

cromwellian replied:
Lazyiness, I didn't get around to writing a helper function to check equality yet.

piotrswigon commented on javatests/arcs/core/data/expression/TypeCheckerTest.kt:
```diff
@@ -0,0 +1,300 @@
+      set("x", IntType)
+      set("y", DoubleType)
+    }.build()
+    checkAssignableTo(query<Number>("x"), IntType, scope, scope)
```
put emptyScope as the ctx ?

i.e.
`checkAssignableTo(query<Number>("x"), IntType, emptyScope, scope)`

piotrswigon commented on javatests/arcs/core/data/expression/TypeCheckerTest.kt:
```diff
@@ -0,0 +1,300 @@
+
+    checkAssignableTo(
+      PaxelParser.parse("from p in numbers select first(words)"),
+      SeqType(UnionType(TextType, NullType)),
```
Do you think representing nullable types as a Union(Type, Null) is a good idea? I'm genuinely not sure...

cromwellian replied:
Technically all types are sets, e.g.

boolean = {true, false}
ints = {0, 1, ..., inf}
text = char*

High types are just products fo these with set union or intersection, so there's nothing wrong with using union types for this. We could add a special wrapper like NullType(InferredType) to make a nullable version, but it would have forced a lot of extra checks elsewhere, whereas unions are general purpose. 

We can change it later if there's good reason.

cromwellian commented on java/arcs/core/data/expression/Builders.kt:
```diff
@@ -213,6 +213,31 @@ open class MapScope<V>(
+    if (other !is MapScope<*>) return false
+
+    if (scopeName != other.scopeName) return false
+    for (key in allKeys) {
```
Good catch

cromwellian commented on java/arcs/core/data/expression/Evaluator.kt:
```diff
@@ -164,42 +168,69 @@ sealed class GlobalFunction(val name: String) {
       toSequence<Comparable<Comparable<*>>>(args[0]).min()
+
+    override fun inferredType(evaluator: TypeEvaluator, args: List<InferredType>) =
+      widen(*args.toTypedArray())
```
yeah, I though we would support max(a,b)/min(a,b) later, but I can limit it for now.

cromwellian commented on java/arcs/core/data/expression/Evaluator.kt:
```diff
@@ -164,42 +168,69 @@ sealed class GlobalFunction(val name: String) {
       toSequence<Int>(args[0]).average()
+
+    override fun inferredType(evaluator: TypeEvaluator, args: List<InferredType>) =
+      InferredType.Primitive.DoubleType
```
The return type of Sequence<Int>.average() is Double

I thought we were gonna loosen these aggregate functions to deal with Sequence<Number> instead of forcing int.

cromwellian commented on java/arcs/core/data/expression/TypeEvaluator.kt:
```diff
@@ -0,0 +1,155 @@
+      )
+
+      // Special ops
+      Expression.BinaryOp.IfNull -> expr.left.accept(this, ctx).union(expr.right.accept(this, ctx))
```
We do track it, but it's done with union right now. So for example, we can require FieldExpression qualifier to not be assignable to NullType.  Perhaps in a later refactor we can decide to represent null as an explicit wrapper (NullableType(InferredType))

cromwellian commented:
PTAL. Added lots of tests, nullability checks, input validation checks, etc

Added accumulation of error and warning messages with node responsible.

Later PR will add source-location error reporting.

cromwellian commented:
/gcbrun

piotrswigon commented on javatests/arcs/core/data/expression/TypeCheckerTest.kt:
```diff
@@ -0,0 +1,479 @@
+    )
+
+    checkTypeIs(
+      PaxelParser.parse("from c in company from p in c.people select p.name"),
```
Nice test!

piotrswigon commented on javatests/arcs/core/data/expression/TypeCheckerTest.kt:
```diff
@@ -0,0 +1,479 @@
+      PaxelParser.parse("from p in numbers where 2 + 3 select p"),
+      SeqType(IntType),
+      scope,
+      errors = listOf("from p in numbers\nwhere 2.0 + 3.0 must evaluate to a boolean type" +
```
oh no, you caught a bug that "2" and "3" are parsed as doubles! This means that doing any arithmetic with constants immediately pushes you to doubles, which is not ideal... One could write 2i + 3i, but that's a bit weird, support for imaginary numbers so early on.. ;-)

piotrswigon commented on javatests/arcs/core/data/expression/TypeCheckerTest.kt:
```diff
@@ -0,0 +1,479 @@
+      PaxelParser.parse("from p in numbers orderby numbers select p"),
+      SeqType(IntType),
+      scope,
+      errors = listOf("order by expression numbers must be a primitive type" +
```
You see that wrapping numbers in quotes would produce a more readable message:
`order by expression "numbers" must be a primitive type but was SeqType(PrimitiveType(Int))`

piotrswigon commented on javatests/arcs/core/data/expression/TypeCheckerTest.kt:
```diff
@@ -0,0 +1,479 @@
+    numeric: Boolean = true
+  ) {
+    checkTypeIs(
+      PaxelParser.parse("from p in numbers select $function($arg)"),
```
Why do you wrap every function invocation in the from/select when the function invocation on its own is a perfectly valid Paxel expression?

cromwellian replied:
Just thought it would be good to test the inferencing falling through the qualified expressions.

piotrswigon commented on javatests/arcs/core/data/expression/TypeCheckerTest.kt:
```diff
@@ -0,0 +1,479 @@
+      PaxelParser.parse("from p in words select now()"),
+      SeqType(LongType),
+      scope
+    )
```
e.g. here why don't you do:
```
    checkTypeIs(
      PaxelParser.parse("now()"),
      LongType,
      scope
    )
```
?

cromwellian replied:
added

piotrswigon commented on javatests/arcs/core/data/expression/TypeCheckerTest.kt:
```diff
@@ -0,0 +1,479 @@
+      SeqType(IntType),
+      scope,
+      errors = listOf("order by expression numbers must be a primitive type" +
+                        " but was SeqType(PrimitiveType(Int)).")
```
Nit: I would consider rendering these types in what I consider a friendlier way, e.g. instead of `SeqType(PrimitiveType(Int))` I would maybe print `Sequence<Int>`, up to you really.

For extra extra points, we could even print `UnionType(X, PrimitiveType(NullType))` as `X?`, though that's a bit more work, still I would argue it would be worthwhile.

cromwellian replied:
Did part 1, extra points save for later, because there are tricky cases:

`A | B?`  (e.g. `Int | String?`) is `UnionType(IntType, UnionType(StringType, NullType))`

vs

`A | B | null` `UnionType(IntType, StringType, NullType)` which arguably should happen but I don't want to add extra complex logic in an already sizable PR.

piotrswigon commented on java/arcs/core/data/expression/InferredType.kt:
```diff
@@ -0,0 +1,133 @@
+
+  /** Represents a type corresponding to a scope with variables of various types. */
+  data class ScopeType(val scope: Expression.Scope) : InferredType() {
+    override fun toString() = "ScopeType(${scope.scopeName})"
```
Not worth printing field names and types?

If I wrote for example: `new Foo {abc: "Yay", def: 42}.abd` and the error told me:
`Field "abd" not found on type "Foo {abc: Text, def: Int}"`
that would be more helpful than:
`Field "abd" not found on type "Foo"`.
(Though I would still manage to figure it out with the second error message).

WDYT?

piotrswigon commented on java/arcs/core/data/expression/Evaluator.kt:
```diff
@@ -164,42 +185,137 @@ sealed class GlobalFunction(val name: String) {
+        InferredType.Primitive.NullType,
+        when (val it = args[0]) {
+          is InferredType.SeqType -> it.type
+          else -> it
```
I believe in `checkInnerSequenceType` above you enforce that the only argument is a sequence, so here you can rely on it. So instead you co just do:
```
return InferredType.UnionType(
  InferredType.Primitive.NullType,
  (args[0] as SeqType).type
)
```

piotrswigon reviewed (approved):
Use your best judgement whether to skip addressing some comments.

cromwellian commented on java/arcs/core/data/expression/TypeEvaluator.kt:
```diff
@@ -0,0 +1,155 @@
+  override fun <E, T> visit(expr: Expression.UnaryExpression<E, T>, ctx: Scope): InferredType {
+    return when (expr.op) {
+      is Expression.UnaryOp.Not -> InferredType.Primitive.BooleanType
+      is Expression.UnaryOp.Negate -> expr.expr.accept(this, ctx)
```
Ugh, I did this, but somehow un-did it before commiting.

cromwellian commented on java/arcs/core/data/expression/TypeEvaluator.kt:
```diff
@@ -0,0 +1,230 @@
+        requireOrWarn(expr, inferredType.isAssignableFrom(InferredType.Primitive.NullType)) {
+          "$expr: ${expr.left} is always not null."
+        }
+        inferredType.union(expr.right.accept(this, ctx))
```
great catch!

cromwellian commented on java/arcs/core/data/expression/TypeEvaluator.kt:
```diff
@@ -0,0 +1,230 @@
+      Expression.BinaryOp.IfNull -> {
+        val inferredType = expr.left.accept(this, ctx)
+        requireOrWarn(expr, inferredType.isAssignableFrom(InferredType.Primitive.NullType)) {
+          "$expr: ${expr.left} is always not null."
```
I think I'll hold off on this until I add location markers in the DSL node. I plan to overall the error messages, and I'll fix all of them up there.
